container/rules/docker_toolchains.bzl

```diff
@@ -308,54 +348,79 @@ def _toolchain_container_impl(ctx):
-              'installables_tar' as .deb files.
-      plus:
-        language_layers: a list of language_tool_layer.
-        installables_tar: a tar of debian packages to install in the base image.
-        packages: list of packages to fetch and install in the base image.
-        additional_repos: list of additional debian package repos to use,
-          in sources.list format.
-        keys: list of labels of additional gpg keys to use while downloading
-          packages.
-        installation_cleanup_commands: cleanup commands to run after package
-          installation.
+            # Declare intermediate output file generated by add_apt_key rule.
+            image_with_keys_output_executable = ctx.actions.declare_file(image_with_keys)
+            image_with_keys_output_tarball = ctx.actions.declare_file(image_with_keys + ".tar")
+            image_with_keys_output_layer = ctx.actions.declare_file(image_with_keys + "-layer.tar")
 
-    If 'installables_tar' is specified in the 'toolchain_container' rule, then
-    'packages' or 'installables_tar' specified in any of the 'language_layers'
-    passed to this 'toolchain_container' rule will be ignored.
+            _key.implementation(
+                ctx,
+                name = image_with_keys,
+                image_tar = ctx.files.base[0],
+                keys = keys,
+                output_executable = image_with_keys_output_executable,
+                output_tarball = image_with_keys_output_tarball,
+                output_layer = image_with_keys_output_layer,
+            )
+            download_base = image_with_keys_output_tarball
 
-    Experimental rule.
-    """
+        # Declare intermediate output file generated by download_pkgs rule.
+        download_pkgs_output_executable = ctx.actions.declare_file(ctx.attr.name + "-download_pkgs_output_executable.sh")
+        if download_pkgs_output_tar:
+            download_pkgs_output_tar = ctx.actions.declare_file(download_pkgs_output_tar)
+        else:
+            download_pkgs_output_tar = ctx.actions.declare_file(ctx.attr.name + ".tar")
+        if download_pkgs_output_script:
+            download_pkgs_output_script = ctx.actions.declare_file(download_pkgs_output_script)
+        else:
+            download_pkgs_output_script = ctx.actions.declare_file(ctx.attr.name + ".sh")
 
-    _input_validation(kwargs)
+        # download_pkgs rule consumes 'packages' and 'additional_repos'.
+        _download.implementation(
+            ctx,
+            image_tar = download_base,
+            packages = packages,
+            additional_repos = additional_repos,
+            output_executable = download_pkgs_output_executable,
+            output_tar = download_pkgs_output_tar,
+            output_script = download_pkgs_output_script,
+        )
+        installables_tars.append(download_pkgs_output_tar)
 
-    toolchain_container_(**kwargs)
+    return struct(
+        providers = [
```

Comment:
Who is consuming the information published in this provider?

Reply:
```toolchain_container``` rule will consume the information here.